### PR TITLE
Support scalar division and unary minus for TensorExpressions

### PR DESCRIPTION
**Pull request by macedo22**
## Proposed changes

This PR adds support for the following operations for `TensorExpression`s:
- unary `-`, e.g. `-R(ti_a, ti_b)`
- division by tensor expressions that evaluate to a rank 0 tensor, e.g.:
    - `(...) / R()`
    - `(...) / sqrt(R(ti_a) * S(ti_A))`
    - `(...) / (R(ti_t, ti_t))`

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp]: `not_null` arguments always go first. All the functions below too

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp:154]: this test is failing, though it's not obvious to me why. The math is slightly different in terms of parentheses, but I'm not sure that's the cause since that should only be rounding errors. I'm also not sure we want to allow this type of math. Doesn't this mean `T_{j}{}^{l} S^{j}{}_{l}` is computed for _every_ LHS index instead of just once? It's probably better to require the `/` rhs be `Scalar<DataType>` (or the expression corresponding to a `Scalar<DataType>`). Considering you as the author of the TensorExpressions might not have noticed this huge performance issue, it's probably best to prevent it completely. Given that change, I'm not reviewing code below in anticipation some of it will change :)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/DataStructures/Tensor/Expressions/Product.hpp
Comment (nilsdeppe) [inline src/DataStructures/Tensor/Expressions/Product.hpp]: I think you want to actually do a unary minus on the underlying type. It's generally not a good idea to stack expressions subtly like this. I'm worried it'll have performance consequences. Can you write a TE class that has a get return `-t_.get();`?

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp
Comment (macedo22) [reply tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp:154]: This is a really good point, and I think it's good to also point out that this is an issue that is not unique to this in-progress `/` operation. For example, all of the following would also suffer from unnecessary repeated calculation:
- `R(ti_a) * (S(ti_B) * T(ti_b))`
- `R(ti_a) - G(ti_a) * sqrt(S(ti_B) * T(ti_b))`
- `R(ti_a) * (S(ti_t, ti_t) + T(ti_t))`

I agree with wanting to avoid the performance issue you've pointed out. However, where do we draw the line with what we permit for the denominator? For example, these denominators are all rank 0 TE types that are not rank 0 `TensorAsExpression` (the TE type for single `Tensor`s):
- ` / (R() * S())` : rank 0 `OuterProduct`, Marlo needs to do this operation.
- `/ (R(ti_t, ti_t) + S(ti_T))` : rank 0 `AddSub`
- `/ sqrt(R())` : rank 0 `SquareRoot`
- `/ sqrt(R(ti_t) - S(ti_t))` : rank 0 `SquareRoot`

So I guess:
- This performance problem wouldn't be unique to `/` with rank 0 TEs, so maybe this is an issue to look into solving across the board.
- What is the extent of the types of denominators we would accept, given some of the examples above? I'm not sure where to draw the line to balance ease of use with what the human should be expected to do up front. Something else to maybe consider:
    -  For a case like `T(ti_c) / (R(ti_A, ti_B) * S(ti_a, ti_b))`, I'm guessing that for the for loop way, an author would be asked to write a separate series of loops to compute this denominator before the separate loop for `c`. And if that's the case, then it seems equally reasonable to ask someone to write something like `evaluate(one_over_denominator, 1.0 / (R(ti_A, ti_B) * S(ti_a, ti_b)))`, and then in `evaluate<ti_c>(result, T(ti_c) * over_over_denominator())`

Let me know your thoughts!

Comment (nilsdeppe) [reply tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp:154]: Yea, you're right. There's no good way to safeguard against this. We probably could do something (in the future) like if you have `op(a, b)` and `a` has 2 or more indices and `b` is a "scalar" (in the sense of it's had a bunch of ops done that in the end produce a scalar) that we tell people to avoid the redundant computations. Definitely beyond the scope of this PR, though :) So please leave as-is

**Comment by macedo22** [comment]
@nilsdeppe Fixups are ready for review!

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please squash

**Comment by macedo22** [comment]
@nilsdeppe Fixups have been squashed

**Comment by macedo22** [comment]
Sorry, I forgot about this failing test. I'll see if I can find a way to run on gcc-10 and report back

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (nilsvu) [inline src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Would the inverse of this function also be useful? E.g. to write `1. / lapse()` as an expression. Related, do you intend to add `Pow` as a TE as well?

**Review by nilsvu — COMMENTED** [review]
I suggest you rebase before we merge this, to make sure there are no conflicts with #3539.

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp
Comment (nilsvu) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp]: I think the issue with the failing test is that this distribution includes zero, but you use it to generate `R` which you divide by.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I think `1.0 / lapse()` would be good to include in this PR. I'll push it up as a fixup.

Regarding `pow`, I've been thinking about this as well. @nilsdeppe do you agree that we should also support `pow`? I can submit this in a separate PR.

**Comment by macedo22** [comment]
@nilsleiffischer @nilsdeppe I've added fixups that add support for `double / Scalar<DataType>` and (hopefully) fix the gcc-10 test failure.

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (nilsvu) [inline src/DataStructures/Tensor/Expressions/Divide.hpp:37]: This is how the other TEs currently work, so no change request for this PR, but perhaps we'll have to generalize this a bit in the future e.g. to support tensors of `ComplexDataVector`. I haven't used `Tensor` with anything but `DataVector` and `double` so far, but in principle it can support more data types than the two.

**Review by nilsvu — APPROVED** [review]
LGTM, please squash once @nilsdeppe has had a chance to look.

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp
Comment (nilsvu) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp]: This is also used in the denominator (line 206 below)

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:37]: Agreed. Right now it is assumed that the data type is `double` or `DataVector`

Comment (nilsvu) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:37]: ok seems fine for now 👍

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp
Comment (macedo22) [reply tests/Unit/DataStructures/Tensor/Expressions/Test_Divide.cpp]: Fixed in new fixup!

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: The reason for not having `1.0 / lapse()` (so far) is to avoid people accidentally doing things like `1.0 / lapse() * shift(ti_I)`. The operation is certainly valid so if we have a lot of `1.0 /` cases we can add support. Note that currently you can do `1.0 / get(lapse)` to produce an inverse.

I haven't look at `pow` in a while. It might be supported in the same way that `square` is

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Ok, does this mean you would like me to remove the support for `1.0 / ScalarTE`?

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Wait and see what @nilsleiffischer has to say. We did explicitly not add it for the reason stated above :)

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Looking at `pow` in `src/Utilities/ConstantExpressions.hpp`([here](https://github.com/sxs-collaboration/spectre/blob/761f78545421d1dd6db157cb2e0e751b22d0c650/src/Utilities/ConstantExpressions.hpp#L159)), I do believe it should already work for positive integral powers, but I don't think `pow` will work for negative integral powers without either:
- supporting `double / ScalarTE<DataType>` generally or
- defining a new `pow` overload that returns `NumberAsExpression(double) / pow<-N>(TensorExpression)`, where `N` is a negative number

Separately, I don't think fractional powers will work until an overload of `cmath`'s `pow` is added that accepts `double` powers.

Comment (nilsvu) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I have no strong opinion on whether or not to allow `1.0 / lapse()`. I feel like the user will always manage to write inefficient code if they want to, and these kind of restrictions sometimes get in the way of writing reasonable code (e.g. "I just want to write 1 / lapse^7, why do I have to get() the underline DataVector for that?"). I'd merge the PR as-is, but won't object if @nilsdeppe requests we don't add the `double / ScalarTE` operation for now.

I think it's more important to solve the issue related to `square`, `pow` and contractions where the same sub-expression is evaluated multiple times, **but not in this PR**. Details below.

---

Do I understand correctly that `square(R(ti_i) * S(ti_I))` will evaluate the contraction twice? That seems quite dangerously expensive, and isn't what a user might intuitively expect (the `square` being "outside" the contraction implies it is applied after the contraction). Does the expression actually work, since it effectively reads `R(ti_i) * S(ti_I) * R(ti_i) * S(ti_I)`, which has inconsistent indices?

I don't yet understand well enough how the current implementation handles memory. Is the full TE essentially one Blaze expression, i.e. a series of arithmetic operations involving DataVectors such as `2. * x + y`, per free index? Does Blaze allocate memory for subexpressions like `2. * x` or does it compute the full expression just by using the output buffer? By extension, I'm assuming the current TE implementation allocates no intermediate memory, but computes each free index separately. E.g. `R(ti_i) * S(ti_I) * T(ti_j)` evaluates `R(ti_i) * S(ti_I)` dim-times, right? In this case the sub-expression `R(ti_i) * S(ti_I)` could be buffered in a component of the output, used to fill the other components, and then multiplied by the component of `T`. Should we make it up to the user to manage memory like that and re-use subexpressions by evaluating them explicitly, and perhaps emit a warning or even a static-assert if sub-expressions are repeated? Or can we build some level of optimization into the TEs, similar to Blaze's intra-statement optimization https://bitbucket.org/blaze-lib/blaze/wiki/Intra-Statement%20Optimization? I think for these decisions and future optimizations it's important to work with real-life use cases, so it's super useful that you're applying TEs for the CCZ4 system now. Here's an example from the XCTS system:

```cpp
/// g_ik g_jl R^ij S^kl
void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
                    const gsl::not_null<DataVector*> buffer1,
                    const gsl::not_null<DataVector*> buffer2,
                    const tnsr::II<DataVector, 3>& tensor1,
                    const tnsr::II<DataVector, 3>& tensor2,
                    const tnsr::ii<DataVector, 3>& metric) noexcept {
  get(*result) = 0.;
  for (size_t i = 0; i < 3; ++i) {
    for (size_t j = 0; j < 3; ++j) {
      *buffer1 = metric.get(i, 0) * tensor1.get(0, j);
      *buffer2 = metric.get(j, 0) * tensor2.get(i, 0);
      for (size_t k = 1; k < 3; ++k) {
        *buffer1 += metric.get(i, k) * tensor1.get(k, j);
        *buffer2 += metric.get(j, k) * tensor2.get(i, k);
      }
      get(*result) += *buffer1 * *buffer2;
    }
  }
}
```

I would very much like to write it as a TE, but will it be as efficient as this implementation? Note that this function actually constitutes a significant fraction of the runtime for a binary black-hole initial-data solve!

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: okay, since the work is already done for `double / Scalar` let's leave it in. I agree that people will inevitably write inefficient code...

The optimization you're looking for in the traces is different from what Blaze does. Blaze commutes operations but doesn't create temporaries. The only way I can think of simplifying the example you gave is if TEs can allocate in the middle, which I think opens up a huge can of bugs + performance issues. You can write it as 2 exps:
```cpp
evaluate<>(scalar, S(ti_I) * R(ti_i));
evaluate<ti_j>(vector, scalar() * T(ti_j));
```
I prefer this level of control since it makes it clear where the allocations are happening. It's slightly less convenient, but I think still much better than the for loops :)

**Review by nilsdeppe — COMMENTED** [review]
go ahead and squash. I have faith that Nils can find the one line change post-squash :)

**Review by nilsdeppe — COMMENTED** [review]
Also please remove all `noexcept` while squashing, otherwise you'll have to rebase again

**Comment by macedo22** [comment]
For the `noexcept`, just remove any in the new code introduced in this PR, or also add another commit that removes `noexcept` from TEs and their tests completely?

**Comment by nilsdeppe** [comment]
this PR should not introduce any new `noexcept`s. Do not remove any existing

**Comment by macedo22** [comment]
@nilsdeppe @nilsleiffischer Ready for re-review after squashing + rebasing + removing `noexcept`

**Review by nilsdeppe — APPROVED** [review]

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I'm just commenting to say yes, `square(R(ti_i) * S(ti_I))` will evaluate the inner product twice and yes, `R(ti_i) * S(ti_I) * T(ti_j)` evaluates `R(ti_i) * S(ti_I)` dim-times because the expression tree has no intermediates (basically is a long blaze expression for `DataVector`) and recurses through the full tree for each unique LHS multi-index. I think I agree with both of you that avoiding this would involve the need for buffers to store intermediate values.

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: And for `pow`, `cube`, and `square`, I think the repeated computation in something like `square(R(ti_I, ti_i))` could be avoided by defining a new overload for `square`, `cube`, and `pow` for TEs and define a `Pow` class (like I think Nils Fischer mentioned), and then that `Pow::get` could just use the cmath `pow`OR the `ConstantExpressions.hpp` `pow` on the result of the contraction. The latter would only currently work for integral powers, though, I think.

Let me know if this is something we're interested in or not. If so, then would we want positive, negative, and fractional powers all supported?

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Because of the implicit memory allocations I mentioned, I think this would be a bad idea to implement.

**Review by nilsvu — APPROVED** [review]

File: src/DataStructures/Tensor/Expressions/Divide.hpp
Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I'm not sure I follow, so I'm going to include here a rough idea of what I was thinking:
```
#include <cmath>

template <typename T>
struct Pow {
  ...

  SPECTRE_ALWAYS_INLINE decltype(auto) get(
      const std::array<size_t, num_tensor_indices>& multi_index) const {
    return pow(t_.get(multi_index), power_);
  }

private:
  T t_;
  const double power_;
};

template <typename T, typename X, typename Symm, typename IndexList,
          typename... Args>
SPECTRE_ALWAYS_INLINE auto pow(
    const TensorExpression<T, X, Symm, IndexList, tmpl::list<Args...>>& t,
    const double power) {
  return TensorExpressions::Pow<T, Args...>(~t, power);
}
```

This doesn't fix repeated the repeated calculations over `j` in something like `R(ti_j) * pow(S(ti_I) * T(ti_i), 4.0)`, as we still compute `pow(S(ti_I) * T(ti_i), 4.0)` `j` times instead of once. However, I think this *does* fix repeated calculations in `pow(S(ti_I) * T(ti_i), 4)` that would happen if we kept things as is and used `ConstantExpressions.hpp`'s `pow` ([here](https://github.com/sxs-collaboration/spectre/blob/761f78545421d1dd6db157cb2e0e751b22d0c650/src/Utilities/ConstantExpressions.hpp#L159)), as this latter `pow` will compute the contraction `4` times and multiply them together instead of computing once and raising to the 4th power. In addition, this latter `pow` only accepts integral powers.

I don't see any implicit allocations / temporary buffers in a setup like this, but please let me know if I'm missing something!

Comment (nilsvu) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I think you're right that your suggestion avoids an intermediate allocation. It should call into `blaze::pow` (https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Operations#!pow). Then it also supports floating-point powers or indeed other vectors / ScalarTE (component-wise powers). The only issue I see is how to avoid that people use the constexpr `pow`. I suppose we could add a TE overload, but that's currently not ideal because Utilities can't depend on DataStructures. @nilsdeppe?

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Yea, but then people should just use `pow(a, b)` instead of `pow<b>(a)` if that's what they want.

Floating point powers are slower to compute than integral powers. SpEC specifically has an `ipow` for this, though this might have change recently :shrug:

I think the way to avoid recomputing things here would be to prohibit `square`, `cube`, `pow<>` with anything that isn't a `Scalar<DataVector>` TE. That is, if you try `square(S(ti_I) * T(ti_i))` you get a "this function is deleted". C++ has the `= delete` syntax that might work here (I haven't looked at it in a few years so I don't remember all the rules)

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: @nilsdeppe Ok, makes sense regarding the floating points. I think the only hangup I have here is that this sounds like the same issue we discussed in this comment thread and [here](https://github.com/sxs-collaboration/spectre/pull/3549#discussion_r716913951) for scalar division, where the repeated calculation due to "bad TE writing" isn't unique to the `pow` operation (of course, while the issue exists elsewhere, it doesn't mean we can't "remove" it here). I think the use case I'm thinking about is similar to what Nils Fischer said above, if you just wanted to be able to do this, where `pow` is a new overload that doesn't do the contraction `4` times:
```
TensorExpressions::evaluate(scalar, pow<4>(S(ti_I) * T(ti_i)));
```
or this:
```
TensorExpressions::evaluate(scalar, pow<4>(S(ti_t) + T(ti_t, ti_t)));
```

so then people have the ability, but we enforce separate `evaluate` calls like [your example](https://github.com/sxs-collaboration/spectre/pull/3549#discussion_r720591179) here if someone tries to do something like this, which we don't want:
```
TensorExpressions::evaluate<ti_j>(result, R(ti_j) * pow<4>(S(ti_I) * T(ti_i)));
```

Do you think the added flexibility warrants supporting this or is it more of a cost than it's worth?

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I don't think special casing `pow<>` to behave as `pow()` is a good idea. That's sneaking things around behind the user's back. In the end all this boils down to that we have no good performance measurements so it's all just hypothesizing. It seems like my idea of disabling cases that are always suboptimal (eg `pow<N>(anything not Scalar<DataVector>)`) is not desired so then we just have to police it in code review.

Comment (nilsvu) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: No I like that idea @nilsdeppe!

Comment (nilsvu) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Do I understand correctly that your suggestion is this:

- Disable `pow<>(TE)`, so users don't accidentally repeat costly tensor expressions. Enable only `pow<>(TensorAsExpression<scalar stuff>)` so people _can_ write `square(lapse())` to get a TE, but _cannot_ write `square(R(ti_i) * S(ti_I))`.
- Add @macedo22's suggestion for `pow(TE, double)` above (https://github.com/sxs-collaboration/spectre/pull/3549#discussion_r721870784).

Please correct this list if I'm misrepresenting your suggestion.

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Yes, that's what I think we should do.

Comment (wthrowe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: Does this problem also occur with Blaze expressions?

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: @nilsleiffischer @nilsdeppe Sounds good :+1: 

@wthrowe Just acknowledging that I see your question, but I don't know the answer. So, I will let one of the Nils take that one!

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: @wthrowe doing:

```cpp
DataVector t{};
td<decltype(pow<4>(t))> aoeu;
```
produces a series of nested mults. We should probably look to see if something like SpEC's `ipow` actually improves matters and use that instead. I also expect that `pow<2>` a mult will win but for sufficiently big powers the `(i)pow` will win

Comment (macedo22) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: @nilsdeppe So then `DataVector` expressions do have the same issue as the TEs in that if you did this:
```
DataVector s{1, 2, 3};
DataVector t{4, 5, 6};
DataVector sum = pow<4>(s + t);
```
`s + t` will be computed 4 times? In my earlier comments in this thread, I was speaking on the assumption that this repeated calculation did not occur with `DataVector`s. So if this _does_ happen, then I take back when I said that the snippet I presented would fix the repeated nested operations.

Comment (nilsdeppe) [reply src/DataStructures/Tensor/Expressions/Divide.hpp:119]: I expect so, but you can check this with `TypeDisplayer` to see what the expression tree looks like in Blaze language

**Pull request merged**
